runtime/minor_gc.c

```diff
@@ -716,7 +709,14 @@ static void caml_stw_empty_minor_heap_no_major_slice(caml_domain_state* domain,
 
   CAML_EV_BEGIN(EV_MINOR_CLEAR);
   caml_gc_log("running stw empty_minor_heap_domain_clear");
-  caml_empty_minor_heap_domain_clear(domain, 0);
+  caml_empty_minor_heap_domain_clear(domain);
+#ifdef DEBUG
+  {
+    uintnat* p = ((uintnat*)(domain->young_end)) - minor_words_used;
+    for (int i=0; i<minor_words_used; ++i) p[i] = Debug_free_minor;
+  }
```

Comment:
This looks like an arguably more convoluted way of doing

```c
for (uintnat *p = (uintnat*)domain->young_ptr; p < (uintnat*)domain->young_end; ++p)
```

Is there a benefit to your definition (maybe the values involved change during some function call earlier?), or is it just a preference?

Note: the code of `caml_empty_minor_heaps_promote` has:

```c
  uintnat minor_allocated_bytes = (uintnat)young_end - (uintnat)young_ptr;
```

I believe that you are right and that code is wrong, we are counting words, not bytes. Am I missing something?

Reply:
(Another question: why not keep this loop in the `caml_empty_minor_heaps_promote` as before? )

Reply:
The loop setting `Debug_free_minor` must take care to not use `domain->young_ptr` after it has been updated in `caml_empty_minor_heaps_promote`. So we could have:
```C
for (uintnat *p = initial_young_ptr;
     p < (uintnat*)domain->young_end;
     ++p)
```

I agree that `minor_allocated_bytes` in `caml_empty_minor_heaps_promote` is poor and needs fixing. 

On moving things in `caml_empty_minor_heaps_promote`, I tread carefully because the function will run in parallel across all domains. We must not destroy values that _another domain_ may promote until all domains have finished their minor collections. We know we can rewrite all the minor heaps after the barrier in the `EV_MINOR_LEAVE_BARRIER` block (~L700) has finished (and not before). 

I'll change the loop to use pointers and fix `minor_allocated_bytes` in `caml_empty_minor_heaps_promote`.

Reply:
I made a mistake in my second message above, apologies: I didn't mean `caml_empty_minor_heaps_promote`, but `caml_empty_minor_heap_domain_clear`, which is the function currently responsible for clearing the minor heap with `Debug_uninit_align`.

Reply:
On closer inspection (aka trying to do it), I now think `minor_allocated_bytes` has no error: 
 - the cast to `(unintnat)` that initializes `minor_allocated_bytes` will correctly give bytes
 - the variable is used as a byte quantity here:
```C
  caml_gc_log ("Minor collection of domain %d completed: %2.0f%% of %u KB live",
               domain->id,
               100.0 * (double)st.live_bytes / (double)minor_allocated_bytes,
               (unsigned)(minor_allocated_bytes + 512)/1024);
```
- and converted to a word quantity here:
```C
  domain->stat_minor_words += Wsize_bsize (minor_allocated_bytes);
```

I have left it as is rather than introduce code motion. 

The loop is changed to pointers in [79cfe30](https://github.com/ocaml/ocaml/pull/11023/commits/79cfe300cb4294ea78d5466277cab0ebddd38233)

Reply:
Your explanation that `_promote` will reset `young_ptr` and `young_limit` (but not `start` and `end`) also explains why we cannot easily keep the cleanup in `_clear`. (I think it would still make sense to do it, by passing `initial_young_ptr` as an argument to that function, but the current approach is also okay.)

Reply:
Note: there is a *lot* of stuff in `caml_empty_minor_heap_promote`! I wish this was split in smaller functions to make it easier to digest.

Reply:
> I made a mistake in my second message above, apologies: I didn't mean caml_empty_minor_heaps_promote, but caml_empty_minor_heap_domain_clear, which is the function currently responsible for clearing the minor heap with Debug_uninit_align.

So for this, we would need to pass in the `initial_young_ptr` to `caml_empty_minor_heap_domain_clear` and use it only in DEBUG. I made the call that placing the change in one place was easier to understand vs saving the pointer and passing to the function which is then only used in DEBUG.